morpheus/controllers/milvus_vector_db_controller.py

```diff
@@ -0,0 +1,185 @@
+        schema_fields_conf = schema_conf.get("schema_fields")
+
+        if not self.has_store_object(name):
+
+            if len(schema_fields_conf) == 0:
+                raise ValueError("Cannot create collection as provided empty schema_fields configuration")
+
+            schema_fields = [self._create_schema_field(field_conf=field_conf) for field_conf in schema_fields_conf]
+
+            schema = CollectionSchema(fields=schema_fields,
+                                      auto_id=schema_conf.get("auto_id", False),
+                                      description=schema_conf.get("description", ""))
+
+            collection = Collection(name=name,
+                                    schema=schema,
+                                    using=self._alias,
+                                    shards_num=collection_conf.get("shards", 2),
+                                    consistency_level=collection_conf.get("consistency_level", "Strong"))
+
+            if partition_conf:
+                # Iterate over each partition configuration
+                for part in partition_conf:
+                    collection.create_partition(part["name"], description=part.get("description", ""))
+            if index_conf:
+                self._create_index(collection=collection,
+                                   field_name=index_conf["field_name"],
+                                   index_params=index_conf["index_params"])
+
+    @with_mutex("_mutex")
+    def insert(self, name, data, **kwargs):
+
+        collection_conf = kwargs.get("collection_conf", {})
+        partition_name = collection_conf.get("partition_name", "_default")
+
+        # TODO (Bhargav): Load input data from a file
+        if isinstance(data, list):
+            if not self.has_store_object(name):
+                raise ValueError(f"Collection {name} doesn't exist.")
+            collection = Collection(name=name)
+            collection.insert(data, partition_name=partition_name)
+            collection.flush()
+
+        elif isinstance(data, pd.DataFrame):
+
+            index_conf = collection_conf.get("index_conf", None)
+            params = collection_conf.get("params", {})
+
+            collection, ins_res = Collection.construct_from_dataframe(
+                name,
+                data,
+                primary_field=collection_conf["primary_field"],
+                auto_id=collection_conf.get("auto_id", False),
+                description=collection_conf.get("description", ""),
+                partition_name=partition_name,
+                **params
+            )
+
+            if index_conf:
+                self._create_index(collection=collection,
+                                   field_name=index_conf["field_name"],
+                                   index_params=index_conf["index_params"])
+
+            collection.flush()
+        else:
+            raise ValueError("Unsupported data type for insertion.")
+
+    @with_mutex("_mutex")
```

Comment:
I'm not sure if we need these mutex at this level. Do you know if the milvus functions are thread-safe?

Reply:
To the best of my knowledge, using a handler instance, we can load collections into memory and unload them. A collection that has been loaded by connection_1 can also be unloaded by connection_2. I need to simulate a scenario where a query initiated by connection_1 is in progress. If connection_2 attempts to release the collection, I want to investigate what happens in this scenario. Need to make sure, will it wait for the other call to finish, or will it disrupt the execution of the query initiated by connection_1?

Reply:
I've changed the mutex level from function to collection level